fix(receipts): Prevent double message send for received receipt

anthonybilinski opened the pull request:
Fixes #2726
Register for receipt handling only once, log if receipt is received that is unexpected, synchronously add receipt of a sent message to map so that the peer can't answer before we are expecting the receipt.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qtox/qtox/4607)
<!-- Reviewable:end -->

sudden6 commented:
Reviewed 1 of 3 files at r1.
Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/persistence/history.cpp, line 237 at r1](https://reviewable.io:443/reviews/qtox/qtox/4607#-KsoIQhFXqLeN3r2eNPu:-KsoIQhFXqLeN3r2eNPv:b-uo31xl) ([raw file](https://github.com/qtox/qtox/blob/d50999556d182b9ffe7b044b14c0722353a82159/src/persistence/history.cpp#L237)):*
> ```cpp
>     }
> 
>     db->execNow(generateNewMessageQueries(friendPk, message, sender, time, isSent, dispName,
> ```

Is there any noticeable performance impact with this synchronous insert? If so maybe we can add a in memory queue for receipts? (in another PR though)

---

*[src/persistence/offlinemsgengine.cpp, line 68 at r1](https://reviewable.io:443/reviews/qtox/qtox/4607#-KsoI2acalBssvUt8-mp:-KsoI2acalBssvUt8-mq:b-36j4mh) ([raw file](https://github.com/qtox/qtox/blob/d50999556d182b9ffe7b044b14c0722353a82159/src/persistence/offlinemsgengine.cpp#L68)):*
> ```cpp
>         receipts.erase(it);
>     } else {
>     	qWarning() << "Received receipt for unknown message";
> ```

Whitespace error, use spaces intead of tab

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4607)*
<!-- Sent from Reviewable.io -->

anthonybilinski commented:
There is a noticeable performance impact, about 200 ms in normal use. It's pretty bad if you spam aggressively while disk is loaded. Working on a small change to move it back to async now, which I can open in a new PR for.

Whitespace fixed.

anthonybilinski commented:
I'll just update this one and keep it async, fix is working

Diadlo commented:
Reviewed 2 of 3 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/persistence/history.cpp, line 237 at r1](https://reviewable.io:443/reviews/qtox/qtox/4607#-KsoIQhFXqLeN3r2eNPu:-Ksor7pQQwKLr2ibXCEe:bm37aik) ([raw file](https://github.com/qtox/qtox/blob/d50999556d182b9ffe7b044b14c0722353a82159/src/persistence/history.cpp#L237)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

Is there any noticeable performance impact with this synchronous insert? If so maybe we can add a in memory queue for receipts? (in another PR though)
</blockquote></details>

Please, add `TODO` comment, to make working asynchronous system

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4607)*
<!-- Sent from Reviewable.io -->

anthonybilinski commented:
Turns out my fix to make it async wasn't working so well and was messing up history.
I'll add the TODO and submit a PR whenever the async change is done.

anthonybilinski commented:
Created https://github.com/qTox/qTox/issues/4608 to address changing it back to async, have a "mostly" working implementation that segfaults on GUI update, looking for feedback :+1:

sudden6 commented:
Seems like you accidentally pushed an additional commit?

anthonybilinski commented:
Whoops, fixed

anthonybilinski commented:
Added async functionality back, safe with any amount of spamming, no lag: https://github.com/qTox/qTox/pull/4615

anthonybilinski commented:
Moved async into this PR in single commit since both address the same original issue

sudden6 commented:
Reviewed 2 of 5 files at r3.
Review status: 2 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/persistence/offlinemsgengine.cpp, line 137 at r3](https://reviewable.io:443/reviews/qtox/qtox/4607#-KsxKywm6hE9nwJYW74b:-KsxKywnXoxci4-sGErA:b-w2yu9z) ([raw file](https://github.com/qtox/qtox/blob/c8779f998665f6f7a37cf6fbbce5dc7480c35aac/src/persistence/offlinemsgengine.cpp#L137)):*
> ```cpp
> }
> 
> void OfflineMsgEngine::processReceipt_(int receiptId)
> ```

Why the underscore in the name of the function? Please change it to something else.

---

*[src/persistence/offlinemsgengine.cpp, line 148 at r3](https://reviewable.io:443/reviews/qtox/qtox/4607#-Kt1x4CfmBQiA60SSKON:-Kt1x4CgjWvdTlfZpgGZ:bwi1746) ([raw file](https://github.com/qtox/qtox/blob/c8779f998665f6f7a37cf6fbbce5dc7480c35aac/src/persistence/offlinemsgengine.cpp#L148)):*
> ```cpp
>     // could either be called from GUI thread when we receive the receipt, or from db process thread when our message
>     // is inserted into history. Can't assume it's safe to update GUI, so need to force GUI thread to update timestamp
>     emit setTimestamp(f->getId(), receiptId);
> ```

AFAICT this is only needed to ensure `updateTimestamp()` is only called from one thread, maybe we can use a construct like https://github.com/qTox/qTox/blob/master/src/core/coreav.cpp#L131 ? Not sure if that's a good idea though, @Diadlo do you have any ideas about this?

I thin this can be made to work without passing a signal to the chatform and then calling back into offlinemsgengine, which would be much cleaner.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4607)*
<!-- Sent from Reviewable.io -->

Talkless commented:
Review status: 2 of 5 files reviewed at latest revision, 11 unresolved discussions.

---

*[src/persistence/offlinemsgengine.cpp, line 77 at r3](https://reviewable.io:443/reviews/qtox/qtox/4607#-Kt1xwiLTAmgAe506zBz:-Kt1xwiLTAmgAe506zC-:b-y9okj9) ([raw file](https://github.com/qtox/qtox/blob/c8779f998665f6f7a37cf6fbbce5dc7480c35aac/src/persistence/offlinemsgengine.cpp#L77)):*
> ```cpp
>     }
>     receipts[receipt].rowId = messageID;
>     receipts[receipt].bRowValid = true;
> ```

This implementation (same in `OfflineMsgEngine::dischargeReceipt()` uses quite a lot of lookups `[receipt]` repeatedly. I'd suggest to refactor for using iterators to something like this:
```
auto it = receipts.find(receipt);
if (it == receipts.end()) {
   it = receipts.insert(receipt, Receipt());
} else if (it.value().bRow... ) {
   ...
}
Receipt& r = it.value();
r.rowId = ...
r.bRowValid = ...
```

---

*[src/persistence/offlinemsgengine.cpp, line 130 at r3](https://reviewable.io:443/reviews/qtox/qtox/4607#-Kt2-YUvlWBoTZKyFt6q:-Kt2-YUvlWBoTZKyFt6r:bpyzasd) ([raw file](https://github.com/qtox/qtox/blob/c8779f998665f6f7a37cf6fbbce5dc7480c35aac/src/persistence/offlinemsgengine.cpp#L130)):*
> ```cpp
>     QMutexLocker ml(&mutex);
> 
>     auto receipt = receipts.find(receiptId);
> ```

const auto receipt, as iterator is not changed later.
Also, could bet .constFind() to return iterator pointing to constant as receipt is not modified as far I can tell.

---

*[src/persistence/offlinemsgengine.cpp, line 131 at r3](https://reviewable.io:443/reviews/qtox/qtox/4607#-Kt2-bCsxKA4PzkMN5f2:-Kt2-bCsxKA4PzkMN5f3:b-p7gu0u) ([raw file](https://github.com/qtox/qtox/blob/c8779f998665f6f7a37cf6fbbce5dc7480c35aac/src/persistence/offlinemsgengine.cpp#L131)):*
> ```cpp
> 
>     auto receipt = receipts.find(receiptId);
>     auto msg = undeliveredMsgs.find(receipt->rowId);
> ```

const auto msg, as iterator is not changed later.

Is it sure that it cannot be `end()` ?

---

*[src/persistence/offlinemsgengine.cpp, line 139 at r3](https://reviewable.io:443/reviews/qtox/qtox/4607#-Kt20zMVLywhKGWf_nR6:-Kt20zMVLywhKGWf_nR7:b9og74g) ([raw file](https://github.com/qtox/qtox/blob/c8779f998665f6f7a37cf6fbbce5dc7480c35aac/src/persistence/offlinemsgengine.cpp#L139)):*
> ```cpp
> void OfflineMsgEngine::processReceipt_(int receiptId)
> {
>     if (!receipts[receiptId].bRecepitReceived || !receipts[receiptId].bRowValid)
> ```

again, look-up performed twice in worst case scenario.

---

*[src/persistence/offlinemsgengine.cpp, line 141 at r3](https://reviewable.io:443/reviews/qtox/qtox/4607#-Kt213Iaa27nxkjOPI6X:-Kt213Iaa27nxkjOPI6Y:b-8e0fji) ([raw file](https://github.com/qtox/qtox/blob/c8779f998665f6f7a37cf6fbbce5dc7480c35aac/src/persistence/offlinemsgengine.cpp#L141)):*
> ```cpp
>     if (!receipts[receiptId].bRecepitReceived || !receipts[receiptId].bRowValid)
>         return;
>     Profile* profile = Nexus::getProfile();
> ```

Profile* **const** profile

---

*[src/persistence/offlinemsgengine.cpp, line 142 at r3](https://reviewable.io:443/reviews/qtox/qtox/4607#-Kt218unvQ5yw5P8eFBL:-Kt218unvQ5yw5P8eFBM:b-ipd6cq) ([raw file](https://github.com/qtox/qtox/blob/c8779f998665f6f7a37cf6fbbce5dc7480c35aac/src/persistence/offlinemsgengine.cpp#L142)):*
> ```cpp
>         return;
>     Profile* profile = Nexus::getProfile();
>     auto receipt = receipts.find(receiptId);
> ```

const auto receipt,
.constFind()

---

*[src/persistence/offlinemsgengine.h, line 57 at r3](https://reviewable.io:443/reviews/qtox/qtox/4607#-Kt21oN34kunBuitAIs7:-Kt21oN34kunBuitAIs8:birf5us) ([raw file](https://github.com/qtox/qtox/blob/c8779f998665f6f7a37cf6fbbce5dc7480c35aac/src/persistence/offlinemsgengine.h#L57)):*
> ```objc
>     struct Receipt
>     {
>         bool bRowValid = false;
> ```

Some new qTox code uses brace-initializers, may be usefull to keep consistency:
bool bRowValid{false};
[ES.23: Prefer the {} initializer syntax](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es23-prefer-the--initializer-syntax)

---

*[src/persistence/offlinemsgengine.h, line 58 at r3](https://reviewable.io:443/reviews/qtox/qtox/4607#-Kt21e9G1FWkxk3fZ83J:-Kt21e9G1FWkxk3fZ83K:bimisvf) ([raw file](https://github.com/qtox/qtox/blob/c8779f998665f6f7a37cf6fbbce5dc7480c35aac/src/persistence/offlinemsgengine.h#L58)):*
> ```objc
>     {
>         bool bRowValid = false;
>         int64_t rowId;
> ```

[ES.20: Always initialize an object](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es20-always-initialize-an-object)
Could be rowId{0} to avoid random-junk-value bugs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4607)*
<!-- Sent from Reviewable.io -->

anthonybilinski committed:
fix(receipts): Prevent double message send for received receipt

Fixes #2726
Register for receipt handling only once, cache receipts that are received before message is writen to history and mark a message as sent once both its receipt has been received and it has been writen to history

anthonybilinski commented:
- Removed underscore from processReceipt, was using it to indicate privateness
- Removed signal, followed linked construct with the difference of using QueuedConnection instead of BlockingQueuedConnection since that method is mutex locked. I also added #include <QCoreApplication> to check if we are on GUI thread.
- Changed to use .find() and .insert() to only do a single lookup. Kept iterator non-const, since with new implementation iterator is reassigned from end if the value is not found in map. Changed some other iterators in processReceipt and updateTimestamp to const, and use constFind.
- Made auto msg const, it should be impossible for it to be end, as method is mutex locked to deliverOfflineMsgs and called only after registerReceipt which inserts the msg, but added additional check/error for robustness
- Made Profile* const
- Changed struct Receipt initialization to use braces, set rowId to 0 (even though it's only checked when bReceiptReceived is set to true).

Diadlo commented:
Please, use reviewable to answer. It make review much easier for us (now here a lot of boxes with unanswered messages). Just click "Done" button below comment, when you make requested change

---

Reviewed 2 of 4 files at r4.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[src/persistence/offlinemsgengine.cpp, line 58 at r4](https://reviewable.io:443/reviews/qtox/qtox/4607#-KtPvcAUWPHwGEzO31Pj:-KtPvcAVdxHyl42bmQdh:b-f7jgsx) ([raw file](https://github.com/qtox/qtox/blob/24a43b7c7308faf35c1b83f8daa993f92b0a96d7/src/persistence/offlinemsgengine.cpp#L58)):*
> ```cpp
>     auto it = receipts.find(receipt);
>     if (it == receipts.end())
>     {
> ```

Brace should be on the same line with `if`. Same for other uses below

---

*[src/persistence/offlinemsgengine.cpp, line 60 at r4](https://reviewable.io:443/reviews/qtox/qtox/4607#-KtPv7uSMJiorhnauVfp:-KtPv7uSMJiorhnauVfq:b-muid4w) ([raw file](https://github.com/qtox/qtox/blob/24a43b7c7308faf35c1b83f8daa993f92b0a96d7/src/persistence/offlinemsgengine.cpp#L60)):*
> ```cpp
>     {
>         it = receipts.insert(receipt, Receipt());
>     } else if ( it->bRecepitReceived ){
> ```

Remove extra spaces and add one before brace. 
I.e. `} else if (it->bRecepitReceived) {`

---

*[src/persistence/offlinemsgengine.cpp, line 163 at r4](https://reviewable.io:443/reviews/qtox/qtox/4607#-KtPwGKk4HcjxAZIm5hX:-KtPwGKk4HcjxAZIm5hY:biuaiyh) ([raw file](https://github.com/qtox/qtox/blob/24a43b7c7308faf35c1b83f8daa993f92b0a96d7/src/persistence/offlinemsgengine.cpp#L163)):*
> ```cpp
>         profile->getHistory()->markAsSent(receipt->rowId);
> 
>     if (QThread::currentThread() == QCoreApplication::instance()->thread())
> ```

`if` should always be with braces

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4607#-:-KtPw_kyXlrN0aL7wbJ0:bqwpqbt)*
<!-- Sent from Reviewable.io -->

anthonybilinski commented:
Review status: 2 of 3 files reviewed at latest revision, 15 unresolved discussions.

---

*[src/persistence/offlinemsgengine.cpp, line 68 at r1](https://reviewable.io:443/reviews/qtox/qtox/4607#-KsoI2acalBssvUt8-mp:-KtQ0WNHT3XIer_Jei4D:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/d50999556d182b9ffe7b044b14c0722353a82159/src/persistence/offlinemsgengine.cpp#L68)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

Whitespace error, use spaces intead of tab
</blockquote></details>

Done.

---

*[src/persistence/offlinemsgengine.cpp, line 77 at r3](https://reviewable.io:443/reviews/qtox/qtox/4607#-Kt1xwiLTAmgAe506zBz:-KtQ0VRvBAj4dfas0qrk:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/c8779f998665f6f7a37cf6fbbce5dc7480c35aac/src/persistence/offlinemsgengine.cpp#L77)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

This implementation (same in `OfflineMsgEngine::dischargeReceipt()` uses quite a lot of lookups `[receipt]` repeatedly. I'd suggest to refactor for using iterators to something like this:
```
auto it = receipts.find(receipt);
if (it == receipts.end()) {
   it = receipts.insert(receipt, Receipt());
} else if (it.value().bRow... ) {
   ...
}
Receipt& r = it.value();
r.rowId = ...
r.bRowValid = ...
```

</blockquote></details>

Done.

---

*[src/persistence/offlinemsgengine.cpp, line 130 at r3](https://reviewable.io:443/reviews/qtox/qtox/4607#-Kt2-YUvlWBoTZKyFt6q:-KtQ0U9eJVIRjHf9U-Z-:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/c8779f998665f6f7a37cf6fbbce5dc7480c35aac/src/persistence/offlinemsgengine.cpp#L130)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

const auto receipt, as iterator is not changed later.
Also, could bet .constFind() to return iterator pointing to constant as receipt is not modified as far I can tell.
</blockquote></details>

Done.

---

*[src/persistence/offlinemsgengine.cpp, line 131 at r3](https://reviewable.io:443/reviews/qtox/qtox/4607#-Kt2-bCsxKA4PzkMN5f2:-KtQ0RLP16zQAIcj1hb-:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/c8779f998665f6f7a37cf6fbbce5dc7480c35aac/src/persistence/offlinemsgengine.cpp#L131)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

const auto msg, as iterator is not changed later.

Is it sure that it cannot be `end()` ?
</blockquote></details>

Done.

---

*[src/persistence/offlinemsgengine.cpp, line 137 at r3](https://reviewable.io:443/reviews/qtox/qtox/4607#-KsxKywm6hE9nwJYW74b:-KtQ0Q5EIHlw-cGj27nB:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/c8779f998665f6f7a37cf6fbbce5dc7480c35aac/src/persistence/offlinemsgengine.cpp#L137)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

Why the underscore in the name of the function? Please change it to something else.
</blockquote></details>

Done.

---

*[src/persistence/offlinemsgengine.cpp, line 139 at r3](https://reviewable.io:443/reviews/qtox/qtox/4607#-Kt20zMVLywhKGWf_nR6:-KtQ0PYsZe3MLjcniPY3:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/c8779f998665f6f7a37cf6fbbce5dc7480c35aac/src/persistence/offlinemsgengine.cpp#L139)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

again, look-up performed twice in worst case scenario.
</blockquote></details>

Done.

---

*[src/persistence/offlinemsgengine.cpp, line 141 at r3](https://reviewable.io:443/reviews/qtox/qtox/4607#-Kt213Iaa27nxkjOPI6X:-KtQ0OOxiQHUxIZ8Y3s_:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/c8779f998665f6f7a37cf6fbbce5dc7480c35aac/src/persistence/offlinemsgengine.cpp#L141)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

Profile* **const** profile
</blockquote></details>

Done.

---

*[src/persistence/offlinemsgengine.cpp, line 142 at r3](https://reviewable.io:443/reviews/qtox/qtox/4607#-Kt218unvQ5yw5P8eFBL:-KtQ0NqoXyysIwMm3TZz:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/c8779f998665f6f7a37cf6fbbce5dc7480c35aac/src/persistence/offlinemsgengine.cpp#L142)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

const auto receipt,
.constFind()
</blockquote></details>

Done.

---

*[src/persistence/offlinemsgengine.cpp, line 148 at r3](https://reviewable.io:443/reviews/qtox/qtox/4607#-Kt1x4CfmBQiA60SSKON:-KtQ0KsbCIaaK8FS9573:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/c8779f998665f6f7a37cf6fbbce5dc7480c35aac/src/persistence/offlinemsgengine.cpp#L148)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

AFAICT this is only needed to ensure `updateTimestamp()` is only called from one thread, maybe we can use a construct like https://github.com/qTox/qTox/blob/master/src/core/coreav.cpp#L131 ? Not sure if that's a good idea though, @Diadlo do you have any ideas about this?

I thin this can be made to work without passing a signal to the chatform and then calling back into offlinemsgengine, which would be much cleaner.
</blockquote></details>

Done.

---

*[src/persistence/offlinemsgengine.cpp, line 58 at r4](https://reviewable.io:443/reviews/qtox/qtox/4607#-KtPvcAUWPHwGEzO31Pj:-KtQ2sDhA85zi4nnM39I:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/24a43b7c7308faf35c1b83f8daa993f92b0a96d7/src/persistence/offlinemsgengine.cpp#L58)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Brace should be on the same line with `if`. Same for other uses below
</blockquote></details>

Done.

---

*[src/persistence/offlinemsgengine.cpp, line 60 at r4](https://reviewable.io:443/reviews/qtox/qtox/4607#-KtPv7uSMJiorhnauVfp:-KtQ2yYN0OWLQBtyScBJ:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/24a43b7c7308faf35c1b83f8daa993f92b0a96d7/src/persistence/offlinemsgengine.cpp#L60)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Remove extra spaces and add one before brace. 
I.e. `} else if (it->bRecepitReceived) {`
</blockquote></details>

Done.

---

*[src/persistence/offlinemsgengine.cpp, line 163 at r4](https://reviewable.io:443/reviews/qtox/qtox/4607#-KtPwGKk4HcjxAZIm5hX:-KtQ30KrekNNmQceiJpk:b-9jxmiu) ([raw file](https://github.com/qtox/qtox/blob/24a43b7c7308faf35c1b83f8daa993f92b0a96d7/src/persistence/offlinemsgengine.cpp#L163)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

`if` should always be with braces
</blockquote></details>

Added braces for if's with content, left without braces for if's that just check a fail condition and return since that was already present.

---

*[src/persistence/offlinemsgengine.h, line 57 at r3](https://reviewable.io:443/reviews/qtox/qtox/4607#-Kt21oN34kunBuitAIs7:-KtQ0IeU3tKlzWqlQCrr:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/c8779f998665f6f7a37cf6fbbce5dc7480c35aac/src/persistence/offlinemsgengine.h#L57)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

Some new qTox code uses brace-initializers, may be usefull to keep consistency:
bool bRowValid{false};
[ES.23: Prefer the {} initializer syntax](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es23-prefer-the--initializer-syntax)

</blockquote></details>

Done.

---

*[src/persistence/offlinemsgengine.h, line 58 at r3](https://reviewable.io:443/reviews/qtox/qtox/4607#-Kt21e9G1FWkxk3fZ83J:-KtQ0G9ap0c2EDs0ExaW:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/c8779f998665f6f7a37cf6fbbce5dc7480c35aac/src/persistence/offlinemsgengine.h#L58)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

[ES.20: Always initialize an object](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es20-always-initialize-an-object)
Could be rowId{0} to avoid random-junk-value bugs.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4607)*
<!-- Sent from Reviewable.io -->

sudden6 commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 1 of 4 files at r4.
Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4607#-:-KtkvabvwcCkfz65LHdq:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

Diadlo commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4607#-:-KuGh6Fu-kAlxW8plNhh:b-jbsf8z)*
<!-- Sent from Reviewable.io -->
